purchase.py

```diff
@@ -177,84 +129,119 @@ def on_change_with_analytic_required(self, name=None):
-                for move in event.moves:
-                    if move.origin == self:
-                        quantity = Uom.compute_qty(move.uom, move.quantity,
-                            self.unit)
-                        if isinstance(event, StockShipmentIn):
-                            pending_quantity += quantity
-                        elif isinstance(event, StockShipmentInReturn):
-                            pending_quantity -= quantity
-            if isinstance(event, AccountInvoice):
-                for line in event.lines:
-                    if line.origin == self:
-                        quantity = Uom.compute_qty(line.unit, line.quantity,
-                            self.unit)
-                        pending_quantity -= quantity
-
-        move_lines = MoveLine.search([
-                ('purchase_line', '=', self),
-                ('account', '=', pending_invoice_account),
-                ('date', '=', date),
-                ])
-        recorded_pending_amount = sum(line.credit - line.debit for line in
-            move_lines)
-        pending_amount = (Currency.compute(self.purchase.currency,
-                Decimal(pending_quantity) * self.unit_price,
-                self.purchase.company.currency) - recorded_pending_amount)
-
-        move_lines = []
-        if pending_amount:
-            move_line = MoveLine()
-            move_line.account = self.product.account_expense_used
-            if move_line.account.party_required:
-                move_line.party = self.purchase.party
-            move_line.purchase_line = self
-            if pending_amount < _ZERO:
-                move_line.credit = abs(pending_amount)
-                move_line.debit = _ZERO
-            else:
-                move_line.debit = pending_amount
-                move_line.credit = _ZERO
-            self._set_analytic_lines(move_line)
-            move_lines.append(move_line)
-
-            move_line = MoveLine()
-            move_line.account = pending_invoice_account
-            if move_line.account.party_required:
-                move_line.party = self.purchase.party
-            move_line.purchase_line = self
-            if pending_amount > _ZERO:
-                move_line.credit = pending_amount
-                move_line.debit = _ZERO
+        quantities = {}
+        for invoice_line in self.invoice_lines:
+            if invoice_line in self.purchase.invoice_lines_ignored:
+                continue
+            if invoice_line.stock_moves:
+                accounting_date = invoice_line.stock_moves[0].effective_date
+            elif invoice_line.invoice:
+                accounting_date = invoice_line.invoice.invoice_date
             else:
-                move_line.debit = abs(pending_amount)
-                move_line.credit = _ZERO
-            move_lines.append(move_line)
-
-        return move_lines
+                accounting_date = self.delivery_date
```

Comment:
delivery_date may be None. So we should fallback to another date, such as purchase_date.